torch/csrc/jit/passes/peephole.cpp

```diff
@@ -15,368 +18,458 @@ static bool mustBeEqual(const c10::optional<T>& a, const c10::optional<T>& b) {
-      // those
-    } else if (
-        node->matches(
-            "aten::mul(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other) ||
-        node->matches(
-            "aten::div(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other)) {
-      // x * 1 == x / 1 == x
-      if (node->get<at::Scalar>(attr::other)->toDouble() == 1) {
-        GRAPH_UPDATE(
-            *node,
-            " (x * 1 == x / 1 == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
-      }
-    } else if (
-        node->matches(
-            "aten::add(Tensor self, Scalar other, Scalar alpha) -> Tensor",
-            /*const_inputs=*/{attr::alpha, attr::other}) ||
-        node->matches(
-            "aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor",
-            /*const_inputs=*/{attr::alpha, attr::other})) {
-      // x + 0 == x - 0 == x
-      if (node->get<at::Scalar>(attr::alpha)->toDouble() == 1 &&
-          node->get<at::Scalar>(attr::other)->toDouble() == 0) {
-        GRAPH_UPDATE(
-            *node,
-            " (x + 0 == x - 0 == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
+
+      if (changed_) {
+        aliasDb_ = torch::make_unique<AliasDb>(graph_);
+        changed_ = false;
       }
-    } else if (
-        node->kind() == aten::Float || node->kind() == aten::Int ||
-        node->kind() == prim::ImplicitTensorToNum) {
-      Node* input_node = node->input()->node();
-      if (input_node->kind() == prim::NumToTensor) {
-        GRAPH_UPDATE(
-            *node,
-            " (x.NumToTensor().ImplicitTensorToNum() == x.NumToTensor()) is replaced with ",
-            node->input()->debugName());
-        node->output()->replaceAllUsesWith(input_node->input());
+
+      if(!aliasDb_->safeToChangeAliasingRelationship(node->inputs(), node->outputs())) {
+        continue;
       }
-    } else if (node->matches("aten::size(Tensor self) -> int[]")) {
-      if (auto ptt = node->input()->type()->cast<TensorType>()) {
-        if (auto sizes = ptt->sizes().concrete_sizes()) {
-          WithInsertPoint guard(node);
-          IValue ival(sizes);
-          auto const_sizes_val = node->owningGraph()->insertConstant(ival);
-          node->output()->replaceAllUsesWith(const_sizes_val);
-        }
+
+      // if (aliasDb_->mayContainAlias(
+      //         node->outputs(), node->owningGraph()->outputs()) &&
+      //     (aliasDb_->mayContainAlias(
+      //          node->inputs(), node->owningGraph()->inputs()) ||
+      //      aliasDb_->mayContainAlias(
+      //          node->inputs(), node->owningGraph()->outputs()))) {
+      //   continue;
+      // }
+
+      // also, if there are writes to either inputs or outputs
+      // the peephole transformations are invalid
+
+      if (aliasDb_->hasWriters(node)) {
```

Comment:
I believe this is redundant with the `safeToChangeAliasingRelationship` check above, so it can be removed